File: 11. Middle element.py
class Node:
    def __init__(self,data):
        self.data=data
        self.next=None

class LinkedList:
    def __init__(self):
        self.head=None
    def append(self,val):
        n=Node(val)
        if not self.head:
            self.head=n
            return
        cur=self.head
        while cur.next:
            cur=cur.next
        cur.next=n
    def middle(self):
        slow=self.head
        fast=self.head
        while fast and fast.next:
            slow=slow.next
            fast=fast.next.next
        return slow
    def print(self):
        cur=self.head; out=[]
        while cur:
            out.append(str(cur.data)); cur=cur.next
        print(' -> '.join(out))

if __name__=='__main__':
    ll=LinkedList(); ll.append(10); ll.append(20); ll.append(30); ll.append(40); ll.append(50)
    ll.print()
    mid=ll.middle()
    if mid: print('Middle:', mid.data)

File: 11. Middle element.cpp
#include <iostream>
using namespace std;
struct Node{int data; Node* next; Node(int d):data(d),next(nullptr){}
};
void append(Node*& h,int v){ Node* n=new Node(v); if(!h){h=n;return;} Node* c=h; while(c->next) c=c->next; c->next=n; }
Node* middle(Node* head){ if(!head) return nullptr; Node* slow=head; Node* fast=head; while(fast && fast->next){ slow=slow->next; fast=fast->next->next; } return slow; }
void print(Node* h){ while(h){ cout<<h->data; if(h->next) cout<<" -> "; h=h->next;} cout<<endl; }
int main(){ Node* head=nullptr; append(head,10); append(head,20); append(head,30); append(head,40); append(head,50); print(head); Node* mid=middle(head); if(mid) cout<<"Middle: "<<mid->data<<endl; return 0; }
